barretenberg/cpp/pil/vm2/data_copy.pil

```diff
@@ -0,0 +1,175 @@
+// |           1 |          0 |          1 |         2 |                1 |        2 |   200 |          6 |
+// |           1 |          0 |          1 |         1 |                1 |        3 |   300 |          7 |
+// +-------------+------------+------------+-----------+------------------+----------+-------+------------+
+
+
+namespace data_copy;
+
+    #[skippable_if]
+    sel_rd_copy + sel_cd_copy = 0;
+
+    // Selectors for calldata_copy and returndata_copy, mutually exclusive given a constrained gadget id
+    pol commit sel_cd_copy;
+    sel_cd_copy * (1 - sel_cd_copy) = 0;
+    pol commit sel_rd_copy;
+    sel_rd_copy * (1 - sel_rd_copy) = 0;
+
+    // Gadget ID is supplied by the execution trace, if non-zero it can be 1 or 2 (instruction spec constrained)
+    // depending on if the operation is calldatacopy or returndatacopy respectively
+    pol commit operation_id;
+    // Bitwise decomposition
+    operation_id = sel_cd_copy + (2 ** 1) * sel_rd_copy;
+    
+    pol SEL = sel_rd_copy + sel_cd_copy;
+
+    // TODO: Since we are memory aware, we need some kind of timestamp
+    pol commit clk;
+
+    // Constrained as inputs from execution trace
+    pol commit src_context_id; // The context that is read from
+    pol commit dst_context_id; // The context that is written to
+
+    pol commit data_copy_size; // This is also the number of writes
+    pol commit data_offset; // This is the offset within the data to start copying
+    pol commit data_addr; // This is the address of the start of the data in parent/child
+    pol commit data_size; // This is the size of the data in parent/child
+
+    pol commit write_addr; // The dst addr, comes from lookup to main trace
+
+    // Control flow management 
+    // These selectors should be implied to be 1 if SEL = 1;
+    pol commit sel_start;
+    sel_start * (1 - sel_start) = 0;
+    // An active row succeeding sel_end has to be a sel_start 
+    #[END_AFTER_START]
+    (sel_cd_copy' + sel_rd_copy') * sel_end * (sel_start' - 1) = 0;
+
+    pol commit sel_end;
+    sel_end * (1 - sel_end) = 0;
+    pol WRITE_CNT_MINUS_ONE = data_copy_size - 1;
+    pol commit next_write_count_inv;
+    // sel_end = 1 IFF data_copy_size - 1 = 0;
+    #[END_WRITE_CONDITION]
+    SEL * (WRITE_CNT_MINUS_ONE * (sel_end * (1 - next_write_count_inv) + next_write_count_inv) - 1 + sel_end) = 0;
+
+    // ===== Writing to dst_context_id =====
+    pol commit one;
+    SEL * (1 - one) = 0;
+    // Data copy size decrements for each row until we end
+    #[DECR_COPY_SIZE]
+    SEL * (1 - sel_end) * (data_copy_size' - data_copy_size + 1) = 0;
+
+    pol commit sel_mem_write;
+    // Regardless of operation, we have to write
+    // TODO check out of bound
+    sel_mem_write = SEL;
+    #[INCR_WRITE_ADDR]
+    (1 - precomputed.first_row) * SEL * (1 - sel_end) * (write_addr' - write_addr - 1) = 0;
+    #[MEM_WRITE]
+    sel_mem_write { clk, write_addr, value, precomputed.zero, one, dst_context_id }
```

Comment:
Could you add comments to the weird cols (precomputed.zero, one). See my other comment.

BTW, you don't need `one`, you can use the selector which will always be one under these conditions `sel_mem_write`.

(check other lookups just in case)